src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -2006,3 +2053,531 @@ ShouldInitiateMetadataSync(bool *lockFailure)
+				ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+								errmsg("Shards of has distributed table  \"%s\" "
+									   "cannot have NULL shard ranges", relationName)));
+			}
+
+			if (ShardIntervalsOverlapWithParams(firstMin, firstMax, secondMin, secondMax,
+												shardIntervalCompareFunction,
+												collationId))
+			{
+				ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+								errmsg("Shard intervals overlap for table \"%s\": "
+									   "%ld and %ld", get_rel_name(relationId),
+									   shardId, shardInterval->shardId)));
+			}
+		}
+
+		table_close(distShardRelation, NoLock);
+	}
+}
+
+
+/*
+ * citus_internal_add_placement_metadata is an internal UDF to
+ * add a row to pg_dist_placement.
+ */
+Datum
+citus_internal_add_placement_metadata(PG_FUNCTION_ARGS)
+{
+	int64 shardId = PG_GETARG_INT64(0);
+	int32 shardState = PG_GETARG_INT32(1);
+	int64 shardLength = PG_GETARG_INT64(2);
+	int32 groupId = PG_GETARG_INT32(3);
+	int64 placementId = PG_GETARG_INT64(4);
+
+	bool missingOk = false;
+	Oid relationId = LookupShardRelationFromCatalog(shardId, missingOk);
+
+	/* only owner of the table is allowed to modify the metadata */
+	EnsureTableOwner(relationId);
+
+	/* we want to serialize all the metadata changes to this table */
+	LockRelationOid(relationId, ShareUpdateExclusiveLock);
+
+	if (!ShouldSkipMetadataChecks())
+	{
+		/* this UDF is not allowed allowed for executing as a separate command */
+		EnsureCoordinatorInitiatedOperation();
+
+		/*
+		 * Even if the table owner is a malicious user, as long as the shard placements
+		 * fit into basic requirements of Citus metadata, the user can only affect its
+		 * own tables. Given that the user is owner of the table, we should allow.
+		 */
+		EnsureShardPlacementMetadataIsSane(relationId, shardId, placementId, shardState,
+										   shardLength, groupId);
+	}
+
+	InsertShardPlacementRow(shardId, placementId, shardState, shardLength, groupId);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * EnsureShardPlacementMetadataIsSane ensures if the input parameters for
+ * the shard placement metadata is sane.
+ */
+static void
+EnsureShardPlacementMetadataIsSane(Oid relationId, int64 shardId, int64 placementId,
+								   int32 shardState, int64 shardLength, int32 groupId)
+{
+	if (shardId <= INVALID_SHARD_ID)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Invalid shard id: %ld", shardId)));
+	}
+	else if (!ShardExists(shardId))
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Shard id does not exists: %ld", shardId)));
+	}
+
+	if (placementId <= INVALID_PLACEMENT_ID)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Shard placement has invalid placement id "
+							   "(%ld) for shard(%ld)", placementId, shardId)));
+	}
+
+	if (!(shardState == SHARD_STATE_ACTIVE ||
+		  shardState == SHARD_STATE_TO_DELETE))
```

Comment:
Let's either remove the shardstate argument from this function, or disallow `SHARD_STATE_TO_DELETE` for now. Since we never pass this in (because we filter active placements), so the only reason for it to be different than `SHARD_STATE_ACTIVE` is a malicious user messing around.